runtime/libtapasco/src/tapasco.hpp

```diff
@@ -0,0 +1,579 @@
+      throw tapasco_error(stringStream.str());
+    }
+
+    Device *device = 0;
+
+    if ((device = tapasco_tlkm_device_alloc(this->tlkm, dev_id)) == 0) {
+      handle_error();
+    }
+
+    return TapascoDevice(device);
+  }
+
+  int num_devices() {
+    // Retrieve the number of devices from the runtime
+    int num_devices = 0;
+    if ((num_devices = tapasco_tlkm_device_len(this->tlkm)) < 0) {
+      handle_error();
+    }
+    return num_devices;
+  }
+
+private:
+  TLKM *tlkm{0};
+};
+
+/**
+ * C++ Wrapper class for TaPaSCo API. Currently wraps a single device.
+ **/
+struct Tapasco {
+  /**
+   * Constructor. Initializes device by default.
+   * Note: Need to check is_ready if using auto-initialization before use.
+   * @param initialize initializes TAPASCO during constructor (may throw
+   *exception!)
+   * @param dev_id device id of this instance (default: 0)
+   **/
+  Tapasco(tlkm_access const access = tlkm_access::TlkmAccessExclusive,
+          DeviceId const dev_id = 0)
+      : driver_internal(),
+        device_internal(driver_internal.allocate_device(dev_id)),
+        default_memory_internal(device_internal.default_memory()) {
+    tapasco_init_logging();
+    this->device_internal.access(access);
+  }
+
+  /**
+   * Destructor. Closes and releases device.
+   **/
+  virtual ~Tapasco() {}
+
+  TapascoDevice &device() { return device_internal; }
+  TapascoDriver &driver() { return driver_internal; }
+  TapascoMemory default_memory() {
+    return this->device_internal.default_memory();
+  }
+
+  template <typename R, typename... Targs>
+  job_future launch(PEId pe_id, RetVal<R> &ret, Targs... args) {
+    JobArgumentList a(this->device_internal.get_device());
+    set_args(a, args...);
+
+    Job *j = this->device_internal.acquire_pe(pe_id);
+    if (j == 0) {
+      handle_error();
+    }
+
+    if (tapasco_job_start(j, a.list()) < 0) {
+      handle_error();
+    }
+
+    return [this, j, &ret, &args...]() {
+      uint64_t ret_val;
+      if (tapasco_job_release(j, &ret_val, true) < 0) {
+        handle_error();
+      }
+      *ret.value = (R)ret_val;
+      return 0;
+    };
+  }
+
+  template <typename... Targs> job_future launch(PEId pe_id, Targs... args) {
+    JobArgumentList a(this->device_internal.get_device());
+    set_args(a, args...);
+
+    Job *j = this->device_internal.acquire_pe(pe_id);
+    if (j == 0) {
+      handle_error();
+    }
+
+    if (tapasco_job_start(j, a.list()) < 0) {
+      handle_error();
+    }
+
+    return [this, j, &args...]() {
+      if (tapasco_job_release(j, 0, true) < 0) {
+        handle_error();
+      }
+      return 0;
+    };
+  }
+
+  /**
+   * Allocates a chunk of len bytes on the device.
+   * @param len size in bytes
+   * @param h output parameter for handle
+   * @param flags device memory allocation flags
+   * @return TAPASCO_SUCCESS if successful, an error code otherwise.
+   **/
+  int alloc(DeviceAddress &h, size_t const len) {
```

Comment:
Here (alloc, free, copy_to, copy_from) the C++-API changed. All methods had an additional flag-argument, free() also had an argument for the size of the memory which should be freed.
I see these arguments are no longer necessary, but maybe we should also keep the methods with the old signatures and just ignore the additional arguments? Then we wouldn't brake old code

Reply:
Yes!